File: leetcode/backtracking/1593. Split a String Into the Max Number of Unique Substrings/py/main.py
class Solution:
    def __init__(self):
      self.res = 0

    def maxUniqueSplit(self, s: str) -> int:
        def dfs(index, pairs):
          if index >= len(s):
            self.res = max(self.res, len(pairs))
            return
          
          for i in range(index + 1, len(s) + 1):
            newStr = s[index:i]
            if newStr in pairs: continue
            pairs.add(newStr)
            dfs(i, pairs)
            pairs.remove(newStr)    

        dfs(0, set())
        
        return self.res

File: leetcode/backtracking/1593. Split a String Into the Max Number of Unique Substrings/js/main.js
// backtrack
const maxUniqueSplit = s => {
  let res = 0;

  dfs(0);

  return res;

  function dfs(index, set = new Set()) {
    if (index >= s.length) {
      res = Math.max(res, set.size);
      return;
    }

    for (let i = index + 1; i <= s.length; i++) {
      let tmp = s.substring(index, i);

      if (!set.has(tmp)) {
        set.add(tmp);
        dfs(i, set);
        set.delete(tmp);
      }
    }
  }
};
